### ISSUE TITLE
Future of `Composition Environment`

### ISSUE THREAD

**Issue by MisterMX** [crossplane/crossplane#3770]
### What problem are you facing?

Support for `EnvironmentConfigs` was added in Crossplane v1.11.0.

### How could Crossplane help solve your problem?

We need a plan to promote it to v1beta1, which may include accepting it as is, changing it, or dropping the feature. Community feedback would help here.
```[tasklist]
### Tracked issues
- [ ] https://github.com/crossplane/crossplane/pull/6016
- [ ] https://github.com/crossplane/docs/pull/824
- [x] https://github.com/crossplane/crossplane/pull/5938
- [x] https://github.com/crossplane-contrib/function-environment-configs/pull/44
- [x] #3965
- [x] https://github.com/crossplane/docs/issues/305
- [ ] #4583
- [ ] #4275
- [ ] https://github.com/crossplane/crossplane/issues/3454: not strictly related, but causing confusion with this feature
- [ ] patch data back to EnvironmentConfigs and have it persistently stored?
- [x] https://github.com/crossplane/crossplane/pull/3981
- [ ] #3723
- [x] #3941
- [x] #3865
- [ ] #3967
- [x] #4080
- [x] #4070
- [ ] #3926
- [ ] #4274
- [ ] #4393
- [ ] https://github.com/crossplane/crossplane/issues/4591
- [ ] https://github.com/crossplane-contrib/function-environment-configs/issues/43
```

**Comment by blakebarnett** [comment]
We are using it successfully as-is.

Some clarity around the patch ordering in documentation and validation that a patch is just not going to work rather than having it fail silently would be great beta criteria. For example if attempting to patch a composed resource and then trying to use that same field elsewhere as if it were a composite field a maybe validatingAdmissionWebhook could deny it as invalid?

**Comment by ytsarev** [comment]
I agree, https://github.com/crossplane/crossplane/issues/3454 behavior should be documented.

Otherwise it is totally unclear why construction like https://github.com/ytsarev/composition-environment-example/blob/main/xvpc/composition.yaml#L36-L38 does not work.

It's not the best UX in general so if there is chance to change the order for ToEnvironment patches, we should try to do it.

@plumbis where are old doc snippets are living now?

E.g. I'm failing to find https://github.com/crossplane/crossplane/pull/3007/files#diff-57c8cd7167236fac0edefa6b72c183bb0fdbd21bdc8ec0a521407eaf436ebc0e in the current state of repos.

It is generally useful to have snippets close to the code as they represent the working samples of the various functionality right in the crossplane repo.

**Comment by ytsarev** [comment]
I've added working(aka not affected by ordering situation) examples of `ToEnvironmentFieldPath` patches to https://github.com/ytsarev/composition-environment-example/commit/2b14f43118905227625980f69c92f08f0524de53 .

**Comment by ytsarev** [comment]
https://github.com/crossplane/docs/issues/380 related concern

**Comment by plumbis** [comment]
> @plumbis where are old doc snippets are living now?

@ytsarev all snippets are still under their versions (v1.10, v1.9). 
https://github.com/crossplane/docs/tree/master/content/v1.10/snippets

They are being removed in v1.11 and future versions. Docs that still reference snippets need to be updated. 

(Happy to discuss why we're removing snippets but it's beyond the scope of this issue)

**Comment by P0t4T0o** [comment]
hi, is it intended that if we select env config by some label in composition and there are multiple configs with that label, only one config is selected per matchLabels list item:
```
 environment:
    environmentConfigs:
      - type: Selector
        selector:
          matchLabels:
            - key: environment-config/app123
              type: Value
              value: select-me
```
would be nice to be able to select and aggregate all configs with matching labels. these could be necessary changes to enable it: https://github.com/P0t4T0o/crossplane/commit/cf8ee8e1f97b658ca1669675d93d9794273e9839

**Comment by ytsarev** [comment]
@P0t4T0o that makes total sense to me and also looks matching [original design](https://github.com/crossplane/crossplane/pull/3008/files#diff-1894e5714a8f815a975c89cc55edb79c1c216dd5c1408c82ab84522d3acc3e83R95-R105) by @MisterMX . 

@P0t4T0o why don't you create PR from the changes you demonstrated? It already looks functional, just some unit test coverage to extend.

**Comment by mithie** [comment]
> We are using it successfully as-is.
> 
> Some clarity around the patch ordering in documentation and validation that a patch is just not going to work rather than having it fail silently would be great beta criteria. For example if attempting to patch a composed resource and then trying to use that same field elsewhere as if it were a composite field maybe a validatingAdmissionWebhook could deny it as invalid?

I agree with that. Documentation should be adjusted accordingly. We were facing the same issue as mentioned above when we were trying to use the `CombineFromComposite` patch type with a value from a status field that was patched by the `FromEnvironmentFieldPath` before. There's a comment on this in the docs of `FromEnvironmentFieldPath`, but it's still confusing since it fails silently. So, I agree with the above. Having some sort of validation would be a good feature.

**Comment by bobh66** [comment]
I added the following comment to #3931 but I'll mention it here as well -  it seems like it should be valid for an EnvironmentConfig Selector to return no results, just as it is valid for a CompositionSelector to return no results and fallback to the defaultComposition. In this case the environment may be constructed of a variety of named and selected EnvironmentConfigs, as well as patches from the Composite, and the absence of one or more selectors may not be a problem, and may even be expected.

I view it as creating a runtime context from a number of sources, where different sources may exist in different scenarios and the absence of a given source is not necessarily an error condition.

I'm also thinking it might be helpful to have a "default" or "init" section under spec.environment which can contain a hard-coded set of "defaults" for the environment, which may or may not be overridden by selected/named environments or patches.  This would also help with the `FromValue` patch type scenario that has been discussed in #3270 

If there is interest in that I can open a separate issue and see about implementation for 1.13

**Comment by blakebarnett** [comment]
> I'm also thinking it might be helpful to have a "default" or "init" section under spec.environment which can contain a hard-coded set of "defaults" for the environment, which may or may not be overridden by selected/named environments or patches. This would also help with the `FromValue` patch type scenario that has been discussed in #3270
> 
> If there is interest in that I can open a separate issue and see about implementation for 1.13

I have wanted exactly this functionality many times. I've ended up creating "cluster-wide" defaults, and then creating free-form status fields to do `CombineToComposite` to make use of the fields that aren't exact fits for each composition. It would be much nicer to be able to set them per-composition in this way.

**Comment by phisco** [comment]
Firstly, I think this issue should be renamed to `Promote Composition Environment to v1beta1` as `EnvironmentConfigs` are just a part of the larger functionality introduced in the original [one-pager](https://github.com/crossplane/crossplane/blob/7c118917b8584a8643708e82b7479c28dffd36b2/design/one-pager-composition-environment.md).

That said, the widespread adoption of this functionality clearly demonstrates that it addresses a real need for users and does so effectively. For these reasons, we aim to promote it to Beta in `v1.15`. This is a good opportunity to incorporate any necessary breaking changes to the current API based on the feedbacks we have received since this feature was introduced in ﻿`v1.11`.

To facilitate this process, I will create a one-pager to collect and start discussing the known pain points and missing functionalities of the current implementation. This will hopefully help us decide the changes required to confidently promote this feature to Beta. [Here](https://github.com/crossplane/crossplane/issues/4583#issuecomment-1785692401) a related comment.

**Comment by phisco** [comment]
FYI: https://github.com/crossplane/crossplane/pull/5061

**Comment by negz** [comment]
@MisterMX @phisco do you think we should reframe this issue to reflect the decision taken in #5061?

> This document proposes not promoting the Composition Environment feature to beta in v1.15, nor setting a timeline for its promotion to beta, investing in enabling Composition Functions to request extra resources allowing them to reimplement the same functionality while exploring other possible approaches.

Here's my understanding of the state of the feature:

* We've made a decision that we don't want to promote it to beta as-is.
* We want to avoid leaving it as alpha as-is, because the longer it exists as-is the harder it is to unwind.
* We plan to remove `spec.environment` and the environment patch types from native P&T (`mode: Resources`).
* We plan to replace it with [function-environment-configs](https://github.com/phisco/function-environment-configs) combined with function-patch-and-transform.
* We want to keep the EnvironmentConfig type but we're not sure where it should live. Possibly a provider?

Is that correct?

**Comment by negz** [comment]
We need to make a decision about the future of this feature.

I don't think it's going to be promoted to beta in its current form. @phisco @jbw976 could we start by removing the EnvConfig specific fields from the Composition type in v1.17?

**Comment by phisco** [comment]
I'd rather deprecate them first for one release, at least. Otherwise we are forcing people using this alpha, but really widely used,feature to also migrate to functions, while native compositions are not deprecated yet. Given that the migration is not seamless, I'd prefer not rushing it.

**Comment by bobh66** [comment]
@phisco @negz any thoughts on promoting `EnvironmentConfig` to v1beta1 so it can be supported when building `Configuration` packages?  Regardless of the composition environment plumbing, the `EnvironmentConfig` resource itself is extremely useful.

**Comment by phisco** [comment]
I'm fine promoting the EnvironmentConfig resource to beta as is, the other option would be to move it out of tree, e.g. provider-environment-configs.
WDYT @negz?

**Comment by negz** [comment]
I maybe have a slight preference for moving it out of tree? Not sure.

My thinking is that it'd be unusual to have a type installed by core that wasn't actually used by core - only by functions. On the other hand it's also a bit unusual to have a "provider" that installs a type but doesn't actually reconcile it, so I don't feel super strongly.

If/when we bump it to beta can we make sure the "old" env config machinery is explicitly marked deprecated. I want to make it clear that we're only promoting the type (for use with functions) and that we still plan to remove the "built in" env config support.

**Comment by bobh66** [comment]
@phisco @negz I'd like to get the `EnvironmentConfig` promoted to beta  in 1.18, or sooner if we move it to `function-environment-configs`.  My concerns with moving it out-of-tree are that you have to install the function in order to get the resource, which introduces another dependency, and it calls into question the ability to include `EnvironmentConfigs` in crossplane `Configuration` packages as described in #3865 

I think the thing that pushes me toward keeping it in-tree is the desire to be able to include it in a `Configuration` package.  That seems to be more Crossplane-ish and I don't know that we would want to support including out-of-tree resources in `Configuration` packages, although including things like `ConfigMaps` might be useful in both `Configuration` and `Function` packages.

Thoughts?  Do we need a new issue for promoting the resource?  Or can we treat the promotion/deprecation as one thing and deal with it here?

**Comment by phisco** [comment]
@bobh66 find keeping this issue.

I'm fine just promoting EnvironmentConfig and leaving the environment machinery to be deprecated with the rest of the native patch and transform composition. I think it would add a lot of useless pain for users to move it out of Crossplane.

**Comment by bobh66** [comment]
Thanks @phisco - is there any additional effort required to deprecate the machinery, or does deprecating the native P&T cover it?

**Comment by negz** [comment]
I feel okay leaving the `EnvironmentConfig` type in-tree. I'd like to actively remove the in-tree code paths though.

By that I mean:

* The `environment` field on the Composition type
* The controller code that selects and fetches environment config
* All the native `FromEnvironment` and `ToEnvironment` patch types

> I think it would add a lot of useless pain for users to move it out of Crossplane.

I'm assuming by "it" here you mean the things I listed above - it's unclear if you meant that or the EnvironmentConfig type.

If so, I empathize, but the flipside is that we must maintain the code for this alpha feature indefinitely. We're committed to maintain most of the in-tree Composition code because it shipped as a GA feature, but EnvironmentConfigs never left alpha.

I know EnvironmentConfigs are popular and widely used, but everyone using them has opted into an alpha feature with plenty of warning that it might be dropped or changed in future. I think it's reasonable to ask them to migrate to the beta iteration of the feature (i.e. using functions).

**Comment by phisco** [comment]
I meant dropping the EnvironmentConfig resource would be painful.

Sounds good to me dropping the machinery around them though, the migration path is as easy as pushing the configuration down to function-environment-config/function-patch-and-transform. If we start making people aware now, it should be fine.

**Comment by phisco** [comment]
As discussed we'll have to:

- [ ] Merge https://github.com/crossplane/crossplane/pull/5938, removing the in-tree code and promoting EnvironmentConfigs.
- [ ] update `crossplane beta convert pipeline-composition` to handle `function-environment-configs` too.
- [x] Document the migration in `function-environment-configs`: https://github.com/crossplane-contrib/function-environment-configs/pull/44
- [ ] Update crossplane docs to reference function-environment-configs in https://docs.crossplane.io/latest/concepts/environment-configs/

**Comment by jbw976** [comment]
@phisco i've taken your task list in the comment above and merged them into the main task list for this epic, along with current status. Thanks for driving this in v1.18! 🙇‍♂️

**Comment by phisco** [comment]
Ok, we finally managed to execute on the plan above, so we can close this!

**TL;DR:** We've promoted the `EnvironmentConfig` resource to `Beta`, but dropped all "in-tree" Composition Environment capabilities, i.e. `spec.environment` from `Compositions`. There is clear documentation and automated tooling (`crossplane beta convert pipeline-composition/composition-environment`) to migrate from that to [function-environment-configs](https://github.com/crossplane-contrib/function-environment-configs). We have almost complete feature-parity, except for `spec.environment.policy.resolve: IfNotPresent`, which only supports `Always` now, but we don't think this will affect users, see [here](https://github.com/crossplane-contrib/function-environment-configs/issues/6) for more details.
